pyresample/bucket/__init__.py

```diff
@@ -0,0 +1,305 @@
+                       adef.area_extent[2] - x_res / 2., x_res)
+    # Orient so that 0-meridian is pointing down
+    y_vect = da.arange(adef.area_extent[3] - y_res / 2.,
+                       adef.area_extent[1] + y_res / 2.,
+                       -y_res)
+
+    result = da.map_blocks(_get_proj_coordinates, lons, lats, x_res, y_res,
+                           prj, new_axis=0,
+                           chunks=(2,) + lons.chunks)
+    proj_x = result[0, :]
+    proj_y = result[1, :]
+
+    # Calculate array indices
+    x_idxs = ((proj_x - np.min(x_vect)) / x_res).astype(np.int)
+    y_idxs = ((np.max(y_vect) - proj_y) / y_res).astype(np.int)
+
+    # Get valid index locations
+    idxs = ((x_idxs >= 0) & (x_idxs < adef.width) &
+            (y_idxs >= 0) & (y_idxs < adef.height))
+    y_idxs = da.where(idxs, y_idxs, -1)
+    x_idxs = da.where(idxs, x_idxs, -1)
+
+    return x_idxs, y_idxs
+
+
+def get_sum_from_bucket_indices(data, x_idxs, y_idxs, target_shape):
+    """Calculate sums for each bin with drop-in-a-bucket resampling.
+
+    Parameters
+    ----------
+    data : Numpy or Dask array
+        Data to be resampled and summed
+    x_idxs : Numpy or Dask array
+        X indices of the target array for each data point
+    y_idxs : Numpy or Dask array
+        Y indices of the target array for each data point
+    target_shape : tuple
+        Shape of the target grid
+
+    Returns
+    -------
+    data : Numpy or Dask array
+        Bin-wise sums in the target grid
+    """
+
+    if isinstance(data, xr.DataArray):
+        data = data.data
+    data = data.ravel()
+    # Remove NaN values from the data when used as weights
+    weights = da.where(np.isnan(data), 0, data)
+
+    # Convert X- and Y-indices to raveled indexing
+    idxs = y_idxs * target_shape[1] + x_idxs
+    # idxs = idxs.ravel()
+
+    out_size = target_shape[0] * target_shape[1]
+
+    # Calculate the sum of the data falling to each bin
+    sums, _ = da.histogram(idxs, bins=out_size, range=(0, out_size),
+                           weights=weights, density=False)
+    return sums.reshape(target_shape)
+
+
+def get_count_from_bucket_indices(x_idxs, y_idxs, target_shape):
+    """Count the number of occurances for each bin using drop-in-a-bucket
+    resampling.
+
+    Parameters
+    ----------
+    x_idxs : Numpy or Dask array
+        X indices of the target array for each data point
+    y_idxs : Numpy or Dask array
+        Y indices of the target array for each data point
+    target_shape : tuple
+        Shape of the target grid
+
+    Returns
+    -------
+    data : Dask array
+        Bin-wise count of hits for each target grid location
+    """
+
+    # Convert X- and Y-indices to raveled index
+    idxs = y_idxs * target_shape[1] + x_idxs
+
+    out_size = target_shape[0] * target_shape[1]
+
+    # Calculate the sum of the data falling to each bin
+    counts, _ = da.histogram(idxs, bins=out_size, range=(0, out_size))
```

Comment:
Can `histogram` be replaced with `bincount` in this case ?

Reply:
Doesn't look like it. In `get_bucket_indices()` I need to mark invalid indices with `-1`, and that doesn't work with `bincount()`. And as the indices need to be integers, using `np.nan` is out... But `histogram()` works :grin: